Improve reporting of errors that occur in event callbacks

kartoFlane opened the pull request:
Errors logged as a result of an error inside an event callback should now include information about where the callback was subscribed, and where the event was dispatched. Addresses #186 

To test this, simply register an event callback that throws an error.

Lemonymous commented:
I have tested the PR using this [ErrorEvent.zip](https://github.com/itb-community/ITB-ModLoader/files/11216939/ErrorEvent.zip) mod.

When using the current mod loader, the following error message is displayed:

```
An event callback failed:  mods/ErrorEvent/scripts/init.lua:12: THIS IS AN ERROR EVENT
```

And, with this PR, the error message displayed is:

```
An event callback failed (dispatched at '...oaderExtensions/mods/modApiExt/scripts/modApiExt.lua:139'): mods/ErrorEvent/scripts/init.lua:14: THIS IS AN ERROR EVENT (subscription created at 'mods/ErrorEvent/scripts/init.lua:13')
```

While this information is helpful, it may be sparse for debugging purposes. A complete `debug.traceback()` is often useful when searching for bugs and understanding the origin of the calls. Additionally, the current layout of the error message on a single long line can make it difficult to read.

If the concern is that including a lot of repeated event callbacks with large amounts of information could slow down the game for regular users, an alternative solution could be to consider [NamesAreHard6600](https://github.com/NamesAreHard6600)'s suggestion of adding more details in development mode.

kartoFlane commented:
I added full tracebacks, and reworked the format so that it's hopefully more readable:
```
An event callback failed: mods/ErrorEvent/scripts/init.lua:12: THIS IS AN ERROR EVENT
- Subscribed at: 
    stack traceback:
        ./scripts/mod_loader/bootstrap/event.lua:135: in function 'subscribe'
        mods/ErrorEvent/scripts/init.lua:11: in function 'nestedFunction3'
        mods/ErrorEvent/scripts/init.lua:17: in function 'nestedFunction2'
        mods/ErrorEvent/scripts/init.lua:21: in function 'nestedFunction1'
        mods/ErrorEvent/scripts/init.lua:25: in function 'init'
        ./scripts/mod_loader/mod_loader.lua:464: in function <./scripts/mod_loader/mod_loader.lua:462>
        [C]: in function 'xpcall'
        ./scripts/mod_loader/mod_loader.lua:461: in function 'initMod'
        ./scripts/mod_loader/mod_loader.lua:157: in function 'init'
        ./scripts/mod_loader/mod_loader.lua:781: in main chunk
        [C]: in function 'require'
        ./scripts/mod_loader/__scripts.lua:42: in main chunk
        [C]: in function 'require'
        scripts/modloader.lua:1: in main chunk
- Dispatched at: 
    stack traceback:
        ./scripts/mod_loader/bootstrap/event.lua:242: in function 'dispatch'
        ...od_loader/extensions/modApiExt/scripts/modApiExt.lua:139: in function 'hook'
        ...mod_loader/extensions/modApiExt/scripts/internal.lua:36: in function <...mod_loader/extensions/modApiExt/scripts/internal.lua:36>
        [C]: in function 'xpcall'
        ...mod_loader/extensions/modApiExt/scripts/internal.lua:34: in function 'firePawnDamagedHooks'
        ...ts/mod_loader/extensions/modApiExt/scripts/alter.lua:98: in function 'trackAndUpdatePawns'
        ...ts/mod_loader/extensions/modApiExt/scripts/alter.lua:767: in function 'fn'
        ./scripts/mod_loader/bootstrap/modApi.lua:167: in function <./scripts/mod_loader/bootstrap/modApi.lua:165>
        (tail call): ?
        [C]: in function 'pcall'
        ./scripts/mod_loader/bootstrap/event.lua:56: in function 'notify'
        ./scripts/mod_loader/bootstrap/event.lua:245: in function 'dispatch'
        ./scripts/mod_loader/bootstrap/modApi.lua:162: in function 'fireMissionUpdateHooks'
        ./scripts/mod_loader/altered/missions.lua:72: in function <./scripts/mod_loader/altered/missions.lua:61>
        (tail call): ?
```

Tracebacks are normally indented using the tab character, which the in-game console does not display. To fix this, I also changed the base logger to replace tab characters with 4 spaces, which makes the tracebacks easier to visually distinguish and read in the middle of other messages in-game.

Lemonymous commented:
It looks good.
Thanks!
